feat: preserve original ref info

marcelltoth opened the pull request:
Needed for https://github.com/stoplightio/json-schema-viewer/issues/125

Adds a new `context` property to `RegularNode` that lets us store auxiliary info, such as the original $ref string we need for JSV125.

LMK if this is a good place to put it, or rather it should go into `meta` or similar?

P0lip commented on src/nodes/RegularNode.ts:
```diff
@@ -31,8 +31,9 @@ export class RegularNode extends BaseNode {
+  public readonly context: Readonly<{ originalRef?: string }>;
 
-  constructor(public readonly fragment: SchemaFragment) {
+  constructor(public readonly fragment: SchemaFragment, context?: { originalRef?: string }) {
```
Hmm, maybe instead of passing `context` we could actually provide the "original fragment"?
This way we could also support combiners that were merged + some other properties that were simply placed next to $ref.

P0lip commented on src/walker/walker.ts:
```diff
@@ -291,7 +292,7 @@ export class Walker extends EventEmitter<WalkerEmitter> {
+          return new RegularNode(merged[0], { originalRef });
         } else {
           const combiner = SchemaCombinerName.OneOf in fragment ? SchemaCombinerName.OneOf : SchemaCombinerName.AnyOf;
           return new RegularNode({
```
so this case essentially.

P0lip commented on src/walker/walker.ts:
```diff
@@ -256,6 +256,7 @@ export class Walker extends EventEmitter<WalkerEmitter> {
   protected processFragment(): SchemaNode {
     const { walkingOptions, path } = this;
     let { fragment } = this;
```
we'd take this fragment as an "original fragment"

```
const { fragment: initialFragment, walkingOptions, path } = this;  // or original, whatever, I suck at naming things.
let { fragment } = this;
```

marcelltoth replied:
Good idea, I guess that would make it more generic

P0lip commented on src/walker/walker.ts:
```diff
@@ -291,7 +291,7 @@ export class Walker extends EventEmitter<WalkerEmitter> {
+          return new RegularNode(merged[0], { originalFragment });
         } else {
           const combiner = SchemaCombinerName.OneOf in fragment ? SchemaCombinerName.OneOf : SchemaCombinerName.AnyOf;
           return new RegularNode({
```
Could we pass this fragment here too?

marcelltoth replied:
Done

P0lip reviewed (approved):
:+1:

marcelltoth commented:
@P0lip it turns out that what I had wasn't enough (when yalced into JSV) because the `oneOf` merging was hiding the fact that one of the items were `$ref`ed. 

I added some logic, which - while still not perfect - I think is an acceptable compromise: We only try to resolve $refs within `oneOf` items when the combiner node doesn't have any properties itself. (It's a pure `{ oneOf: [{}, {}] }` type node.)

This works for me in JSV, but please take another look at the code as I've changed stuff around.

P0lip commented on src/walker/walker.ts:
```diff
@@ -271,7 +274,7 @@ export class Walker extends EventEmitter<WalkerEmitter> {
         } catch (ex) {
           super.emit('error', createMagicError(ex));
-          return new ReferenceNode(fragment, ex?.message ?? 'Unknown resolving error');
+          return new ReferenceNode(fragment, ex?.message ?? 'Unknown   resolving error');
```
```suggestion
          return new ReferenceNode(fragment, ex?.message ?? 'Unknown resolving error');
```
